pensieve/export_json.py

```diff
@@ -0,0 +1,114 @@
+    """Return all blobs in the GCS location with their last modified timestamp."""
+    blobs = storage_client.list_blobs(bucket)
+
+    return {blob.name.replace(".json", ""): blob.updated for blob in blobs}
+
+
+def _export_table(
+    client: bigquery.Client,
+    project_id: str,
+    dataset_id: str,
+    table: str,
+    bucket: str,
+    storage_client: storage.Client,
+):
+    """Export a single table or view to GCS as JSON."""
+    # since views cannot get exported directly, write data into a temporary table
+    job = client.query(
+        f"""
+        SELECT *
+        FROM {dataset_id}.{table}
+    """
+    )
+
+    job.result()
+
+    # get the temporary table results are written to
+    tmp_table = job._properties["configuration"]["query"]["destinationTable"]
+
+    destination_uri = f"gs://{bucket}/{table}.ndjson"
+    dataset_ref = bigquery.DatasetReference(project_id, tmp_table["datasetId"])
+    table_ref = dataset_ref.table(tmp_table["tableId"])
+
+    logging.info(f"Export table {table} to {destination_uri}")
+
+    job_config = bigquery.ExtractJobConfig()
+    job_config.destination_format = "NEWLINE_DELIMITED_JSON"
+    extract_job = client.extract_table(
+        table_ref, destination_uri, location="US", job_config=job_config
+    )
+    extract_job.result()
+
+    # convert ndjson to json
+    _convert_ndjson_to_json(bucket, table, storage_client)
+
+
+def _convert_ndjson_to_json(bucket_name: str, table: str, storage_client: storage.Client):
+    """Converts the provided ndjson file on GCS to json."""
+    ndjson_blob_path = f"gs://{bucket_name}/{table}.ndjson"
+    json_blob_path = f"gs://{bucket_name}/{table}.json"
+
+    logging.info(f"Convert {ndjson_blob_path} to {json_blob_path}")
+
+    # stream from GCS
+    with smart_open.open(ndjson_blob_path) as fin:
+        first_line = True
+
+        with smart_open.open(json_blob_path, "w") as fout:
+            fout.write("[")
+
+            for line in fin:
+                if not first_line:
+                    fout.write(",")
+
+                fout.write(line.replace("\n", ""))
+                first_line = False
+
+            fout.write("]")
+            fout.close()
+            fin.close()
+
+    # delete ndjson file from bucket
+    logging.info(f"Remove file {table}.ndjson")
+    bucket = storage_client.bucket(bucket_name)
+    blob = bucket.blob(f"{table}.ndjson")
+    blob.delete()
+
+
+def export_statistics_tables(project_id: str, dataset_id: str, bucket: str):
+    """Export statistics tables that have been modified or added to GCS as JSON."""
+    bigquery_client = bigquery.Client(project_id)
+    storage_client = storage.Client()
+
+    tables = _get_statistics_tables(bigquery_client, dataset_id)
+    exported_json = _get_gcs_blobs(storage_client, bucket)
+
+    for table, table_updated in tables.items():
+        if table not in exported_json or table_updated > exported_json[table]:
```

Comment:
does this just do a string comparison on the timestamps? parsing them in the _get functions first would feel virtuous

Reply:
It looks like the timestamps are numeric - milliseconds since epoch (given this call `TIMESTAMP_MILLIS()`). As long as they're both in this format, this comparison seems fine to me.

Reply:
fwiw that's backwards; `TIMESTAMP_MILLIS` creates a TIMESTAMP value from a numeric epoch-time representation.

Reply:
The timestamps get returned as `datetime.datetime`

Reply:
That suggests the function signature at https://github.com/mozilla/pensieve/pull/106/files#diff-11973975ebf08e97a7623bf89e189d57R10 needs a buff

Reply:
I'll open a follow-up PR since I just merged this one.

Reply:
Thanks, sorry for the late catch